dsp/modules/llamaindex.py

```diff
@@ -0,0 +1,335 @@
+"""
+this code implements a wrapper around the llama_index library to emulate a dspy llm
+
+this allows the llama_index library to be used in the dspy framework since dspy has limited support for LLMs
+
+This code is a slightly modified copy of dspy/dsp/modules/azure_openai.py
+
+The way this works is simply by creating a dummy openai client that wraps around any llama_index LLM object and implements .complete and .chat
+
+tested with python 3.12
+
+dspy==0.1.4
+dspy-ai==2.4.9
+llama-index==0.10.35
+llama-index-llms-openai==0.1.18
+
+"""
+
+import json
+import logging
+from typing import Any, Literal
+
+from easydict import EasyDict
+from llama_index.core.base.llms.types import ChatMessage
+from llama_index.core.llms import LLM
+
+
+def LlamaIndexOpenAIClientWrapper(llm: LLM):
+    def chat(messages: list[ChatMessage], **kwargs) -> Any:
+        return llm.chat([ChatMessage(**message) for message in messages], **kwargs)
+
+    def complete(prompt: str, **kwargs) -> Any:
+        return llm.complete(prompt, **kwargs)
+
+    client = EasyDict(
+        {
+            'chat': EasyDict({'completions': EasyDict({'create': chat})}),
+            'completion': EasyDict({'create': complete}),
+            'ChatCompletion': EasyDict({'create': chat}),
+            'Completion': EasyDict({'create': complete}),
+        }
+    )
+    return client
+
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO,
+    format='%(message)s',
+    handlers=[logging.FileHandler('azure_openai_usage.log')],
+)
+
+import functools
+import json
+from typing import Any, Literal
+
+import backoff
+import dsp
+import openai
+from dsp.modules.cache_utils import CacheMemory, NotebookCacheMemory, cache_turn_on
+from dsp.modules.lm import LM
+
+try:
+    OPENAI_LEGACY = int(openai.version.__version__[0]) == 0
+except Exception:
+    OPENAI_LEGACY = True
+
+try:
+    import openai.error
+    from openai.openai_object import OpenAIObject
+
+    ERRORS = (
+        openai.error.RateLimitError,
+        openai.error.ServiceUnavailableError,
+        openai.error.APIError,
+    )
+except Exception:
+    ERRORS = (openai.RateLimitError, openai.APIError)
+    OpenAIObject = dict
+
+
+def backoff_hdlr(details):
+    """Handler from https://pypi.org/project/backoff/"""
+    print(
+        'Backing off {wait:0.1f} seconds after {tries} tries ' 'calling function {target} with kwargs ' '{kwargs}'.format(**details),
+    )
+
+
+class DspyLlamaIndexWrapper(LM):
```

Comment:
refactor name to LlamaIndexWrapper. and please add to imports in [lm.py](https://github.com/stanfordnlp/dspy/blob/main/dsp/modules/__init__.py) as done with other modules